### DISCUSSION TITLE
more parameter settings are required for algorithms endpoints

### DISCUSSION THREAD

**Discussion by JinIgarashi** [developmentseed/titiler#760] [Feature Request]
Currently titiler's algorithms api returns parameters with `default`, `title` and `type` only. It would be nice if the api can return more settings for parameters such as `maxValue`, `minValue`. If these settings are available, it is easier to make slider UI for parameters. 

In addition, in default example of algorithms, there is not string parameters. it would be nice if option values can be included in parameter settings (like for making selecting box) if it is string type.

There additional parameter settings can be useful if frontend side want to build a UI dynamically.

**Comment by vincentsarago** [comment]
@JinIgarashi I'm not quite sure to fully understand, to you have an example? 

https://github.com/developmentseed/titiler/blob/70c18b344ef4a8c6e3e83a5bf225cb4047c1bd30/src/titiler/core/titiler/core/algorithm/dem.py#L12-L23
```
curl https://titiler.xyz/algorithms/hillshade | jq
{
  "inputs": {
    "nbands": 1
  },
  "outputs": {
    "nbands": 1,
    "dtype": "uint8",
    "min": null,
    "max": null
  },
  "parameters": {
    "azimuth": {
      "default": 90,
      "title": "Azimuth",
      "type": "integer"
    },
    "angle_altitude": {
      "default": 90.0,
      "title": "Angle Altitude",
      "type": "number"
    },
    "buffer": {
      "default": 3,
      "title": "Buffer",
      "type": "integer"
    }
  }
}
```

everything comes from https://github.com/developmentseed/titiler/blob/main/src/titiler/core/titiler/core/factory.py#L1654C30-L1654C64 (pydantic method) 

```python
HillShade.model_json_schema()["properties"]

{'input_nbands': {'default': 1, 'title': 'Input Nbands', 'type': 'integer'},
 'output_nbands': {'default': 1, 'title': 'Output Nbands', 'type': 'integer'},
 'output_dtype': {'default': 'uint8',
  'title': 'Output Dtype',
  'type': 'string'},
 'output_min': {'anyOf': [{'items': {}, 'type': 'array'}, {'type': 'null'}],
  'default': None,
  'title': 'Output Min'},
 'output_max': {'anyOf': [{'items': {}, 'type': 'array'}, {'type': 'null'}],
  'default': None,
  'title': 'Output Max'},
 'azimuth': {'default': 90, 'title': 'Azimuth', 'type': 'integer'},
 'angle_altitude': {'default': 90.0,
  'title': 'Angle Altitude',
  'type': 'number'},
 'buffer': {'default': 3, 'title': 'Buffer', 'type': 'integer'}}
```

in your custom algorithm, if you see min/max value for a properties using Pydantic `Field` you should get the info in the endpoint 

```python
from titiler.core.algorithm.base import BaseAlgorithm
from pydantic import Field
from rio_tiler.models import ImageData

class Something(BaseAlgorithm):

    # parameters
    yo: int = Field(90, ge=0, lt=100)

    def __call__(self, img: ImageData) -> ImageData:
        return img

Something.model_json_schema()["properties"]

{'input_nbands': {'anyOf': [{'type': 'integer'}, {'type': 'null'}],
  'default': None,
  'title': 'Input Nbands'},
 'output_nbands': {'anyOf': [{'type': 'integer'}, {'type': 'null'}],
  'default': None,
  'title': 'Output Nbands'},
 'output_dtype': {'anyOf': [{'type': 'string'}, {'type': 'null'}],
  'default': None,
  'title': 'Output Dtype'},
 'output_min': {'anyOf': [{'items': {}, 'type': 'array'}, {'type': 'null'}],
  'default': None,
  'title': 'Output Min'},
 'output_max': {'anyOf': [{'items': {}, 'type': 'array'}, {'type': 'null'}],
  'default': None,
  'title': 'Output Max'},
 'yo': {'default': 90,
  'exclusiveMaximum': 100,
  'minimum': 0,
  'title': 'Yo',
  'type': 'integer'}}
```

> In addition, in default example of algorithms, there is not string parameters. it would be nice if option values can be included in parameter settings (like for making selecting box) if it is string type.

Again I'm not sure to get this, but have to use JSON encoded parameters for algorithm because we need `named` parameters and we also need type checking from pydantic

**Reply by JinIgarashi** [reply]
currently, for example, hillshade algorithms returns 

```json
  "parameters": {
    "azimuth": {
      "default": 90,
      "title": "Azimuth",
      "type": "integer"
    },
    "angle_altitude": {
      "default": 90.0,
      "title": "Angle Altitude",
      "type": "number"
    },
    "buffer": {
      "default": 3,
      "title": "Buffer",
      "type": "integer"
    }
  }
```

but for users, we are unable to know what range of values are expected for these parameters. They only have `default` and `type`. It would be nice if titiler also returns `min` and `max`.

for instance, about `azimuth` parameter, we can know the value should be between 0 and 360 (as human being reading), but it is not easy to decide the range of values by programming code to create a UI dynamic. I think now my point is clear to you?

**Reply by vincentsarago** [reply]
Ok oK 👍 

so it's not a general problem, but more a problem with how the Hillshade algorithm class is currently written. Please feel free to open a PR to add the `min` / `max` values for the parameters 🙏

**Reply by JinIgarashi** [reply]
@vincentsarago I created a PR https://github.com/developmentseed/titiler/pull/783 for this
